src/object/srv_obj_migrate.c

```diff
@@ -1766,6 +1860,129 @@ migrate_one_destroy(struct migrate_one *mrone)
 	D_FREE(mrone);
 }
 
+enum {
+	OBJ_ULT = 1,
+	DKEY_ULT = 2,
+};
+
+/* Check if there are enough resource for the migration to proceed. */
+static int
+migrate_system_enter(struct migrate_pool_tls *tls, int tgt_idx)
+{
+	uint32_t tgt_cnt = 0;
+	int	 rc = 0;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+	D_ASSERTF(tgt_idx < dss_tgt_nr, "tgt idx %d tgt nr %u\n", tgt_idx, dss_tgt_nr);
+
+	tgt_cnt = atomic_load(&tls->mpt_obj_ult_cnts[tgt_idx]) +
+		  atomic_load(&tls->mpt_dkey_ult_cnts[tgt_idx]);
+
+	while ((tls->mpt_inflight_max_ult / dss_tgt_nr) <= tgt_cnt) {
+		D_DEBUG(DB_REBUILD, "tgt%d:%u max %u\n",
+			tgt_idx, tgt_cnt, tls->mpt_inflight_max_ult / dss_tgt_nr);
+		ABT_mutex_lock(tls->mpt_inflight_mutex);
+		ABT_cond_wait(tls->mpt_inflight_cond, tls->mpt_inflight_mutex);
+		ABT_mutex_unlock(tls->mpt_inflight_mutex);
+		if (tls->mpt_fini)
+			D_GOTO(out, rc = -DER_SHUTDOWN);
+
+		tgt_cnt = atomic_load(&tls->mpt_obj_ult_cnts[tgt_idx]) +
+			  atomic_load(&tls->mpt_dkey_ult_cnts[tgt_idx]);
+	}
```

Comment:
This kind of "while (check_condition) { wait }" is quite inefficient where there are large amount of waiters.

Do you know how many waiters there could be? If there are at most tens of waiters, I think we can live with it; If there could be hundreds or thousands of waiters, the overhead of the unnecessary cycle of "wakeup -> recheck -> go back to wait" for most waiters will kill the performance badly, to make it worse, lots of contention will be generated by these atomic operations.

Reply:
seems at most one waiter per tgt @wangdi1 ?